In [1]:
import pandas as pd
import numpy as np
import os
from PIL import Image
import shutil

In [65]:
# set the project patch
my_path = "/Users/jiten/Masters/WorkPlace/"
folder_path = "/Users/jiten/Masters/WorkPlace/MRI_MULTI_IMAGE_SEGMENTATION/"
csv_path_SAGIR = os.path.join(folder_path,'boundingInfo','SAGIR bounding boxes.csv')
image_dir = os.path.join(my_path, 'MRI_Image','06192023 SFI renamed')

# set the destinatin fracture and edema folder path
dest_frac_SAGIR = os.path.join(folder_path, 'SAGIR_frac_Images')
dest_edem_SAGIR = os.path.join(folder_path, 'SAGIR_edem_Images')

# create the folders if it do not exist
if not os.path.exists(dest_frac_SAGIR):
    # Create the directory
    os.makedirs(dest_frac_SAGIR)
    print(f"Created folder: {dest_frac_SAGIR}")

if not os.path.exists(dest_edem_SAGIR):
    # Create the directory
    os.makedirs(dest_edem_SAGIR)
    print(f"Created folder: {dest_edem_SAGIR}")

# set the healthy destination folder
healthy_SAGIR = os.path.join(folder_path, 'SAGIR_healthy')

# create the folders if it do not exist
if not os.path.exists(healthy_SAGIR):
    # Create the directory
    os.makedirs(healthy_SAGIR)
    print(f"Created folder: {healthy_SAGIR}")


In [76]:
# read the bounding box csv file
df_bounding = pd.read_csv(csv_path_SAGIR)

# get the image names tha has a bounding info
imgName = df_bounding['image'].unique()

In [67]:
df_bounding

image  original_width  original_height    value_x  \
0     P001 SAGIR_007.jpg             256              256  60.808926   
1     P001 SAGIR_008.jpg             256              256  62.622036   
2     P001 SAGIR_009.jpg             256              256  62.064156   
3     P001 SAGIR_010.jpg             256              256  60.669456   
4     P001 SAGIR_011.jpg             256              256  64.993026   
...                  ...             ...              ...        ...   
1758  P115 SAGIR_016.jpg             256              290  32.583422   
1759  P115 SAGIR_017.jpg             256              290  48.187979   
1760  P115 SAGIR_017.jpg             256              290  42.659238   
1761  P115 SAGIR_017.jpg             256              290  33.192624   
1762  P115 SAGIR_017.jpg             256              290  26.949975   

        value_y  value_width  value_height value_rectanglelabels  
0     38.772664     7.391911     12.412831             ['Edema']  
1     39.051604     7.112971      9.205021             ['Edema']  
2     36.680614    10.739191     12.691771             ['Edema']  
3     36.680614    12.691771     11.576011             ['Edema']  
4     42.538354     7.252441      4.463040          ['Fracture']  
...         ...          ...           ...                   ...  
1758  65.053763     8.678805      3.897849             ['Edema']  
1759  50.627615     3.949834      3.486750          ['Fracture']  
1760  46.311356    16.553175     16.716718             ['Edema']  
1761  62.768817     7.156208      6.989247             ['Edema']  
1762  65.456989     4.872312      6.854839             ['Edema']  

[1763 rows x 8 columns]

In [68]:
df_bounding[df_bounding['image'] == 'P001 SAGIR_007.jpg']['value_rectanglelabels']

0    ['Edema']
Name: value_rectanglelabels, dtype: object

In [69]:
df_bounding[df_bounding['value_rectanglelabels'] == '[\'Edema\']']

image  original_width  original_height    value_x  \
0     P001 SAGIR_007.jpg             256              256  60.808926   
1     P001 SAGIR_008.jpg             256              256  62.622036   
2     P001 SAGIR_009.jpg             256              256  62.064156   
3     P001 SAGIR_010.jpg             256              256  60.669456   
5     P001 SAGIR_011.jpg             256              256  60.111576   
...                  ...             ...              ...        ...   
1757  P115 SAGIR_016.jpg             256              290  38.575154   
1758  P115 SAGIR_016.jpg             256              290  32.583422   
1760  P115 SAGIR_017.jpg             256              290  42.659238   
1761  P115 SAGIR_017.jpg             256              290  33.192624   
1762  P115 SAGIR_017.jpg             256              290  26.949975   

        value_y  value_width  value_height value_rectanglelabels  
0     38.772664     7.391911     12.412831             ['Edema']  
1     39.051604     7.112971      9.205021             ['Edema']  
2     36.680614    10.739191     12.691771             ['Edema']  
3     36.680614    12.691771     11.576011             ['Edema']  
5     39.051604    14.783821     10.320781             ['Edema']  
...         ...          ...           ...                   ...  
1757  46.455887    28.722290     14.880551             ['Edema']  
1758  65.053763     8.678805      3.897849             ['Edema']  
1760  46.311356    16.553175     16.716718             ['Edema']  
1761  62.768817     7.156208      6.989247             ['Edema']  
1762  65.456989     4.872312      6.854839             ['Edema']  

[1430 rows x 8 columns]

In [77]:
print("Total no of Fracture Images having bounding box :::", len(df_bounding[df_bounding['value_rectanglelabels'] == '[\'Fracture\']']))

Total no of Fracture Images having bounding box ::: 333


In [78]:
print("Total no of unique Fracture Images having bounding box :::",len(df_bounding[df_bounding['value_rectanglelabels'] == '[\'Fracture\']']['image'].unique()))

Total no of unique Fracture Images having bounding box ::: 325


In [79]:
print("Total no of Edema Images having bounding box :::", len(df_bounding[df_bounding['value_rectanglelabels'] == '[\'Edema\']']))

Total no of Edema Images having bounding box ::: 1430


In [80]:
print("Total no of unique Edema Images having bounding box :::",len(df_bounding[df_bounding['value_rectanglelabels'] == '[\'Edema\']']['image'].unique()))

Total no of unique Edema Images having bounding box ::: 1015


In [57]:
#df_bounding['value_rectanglelabels'].apply(eval)

**There are total 1015 images some having both edema and fractures and some having only multiple edema and no fracture**

### Lets pull all the Fracture and Edema Images to its repective folder

In [86]:
# we are dealing with SAGIR images
img = 'SAGIR'

def prepareData(imgName, img):
    """
    Function to copy the images from main data folder to
    a different folder for all img images
    """
    
    # initialize the counters to track the number of copies
    unhealthy_counter = 0
    unhealthy_frac_count = 0
    unhealthy_edema_count = 0
    healthy_counter = 0

    print(f"Total unique unhealty images are :::: {len(imgName)}")
    # images having Edema
    edema_img_names = df_bounding[df_bounding['value_rectanglelabels'] == '[\'Edema\']']['image'].unique()
    print(f"Total  unique Images that has edema are ::: {len(edema_img_names)}")
    # images having Fracture
    frac_img_names = df_bounding[df_bounding['value_rectanglelabels'] == '[\'Fracture\']']['image'].unique()
    print(f"Total  unique Images that has fracture are ::: {len(frac_img_names)}")

    print("Start of image seggregation to respective folders......")

    # go through the images and segregate
    for root, dirs, files in os.walk(image_dir):
        for file_name in files:
            # ignore the store file
            if file_name != '.DS_Store':
                file_path = os.path.join(root, file_name)
                # if it is a SAGIR image 
                if img in file_path:
                    # take the unhealty images to differnt folder for training and testing
                    if file_name in imgName:
                        # if the file has edema copy it to the edema folder 
                        # and if it is also fracture copyto the fracture folder
                        if file_name in edema_img_names:
                            shutil.copy(file_path, dest_edem_SAGIR)
                            unhealthy_edema_count += 1
                        if file_name in frac_img_names:
                            shutil.copy(file_path, dest_frac_SAGIR)
                            unhealthy_frac_count += 1
                        unhealthy_counter += 1

                    if file_name not in imgName:
                        # copy only those healthy image slice which has more than 65% of the image as foot
                        h_img = Image.open(file_path)
                        h_img_gray = h_img.convert("L")
                        h_img_array = np.array(h_img_gray)
                        h_size = h_img.size
                        black_pix = np.sum(h_img_array <= 10) # pixel value <=10 considered as black

                        if( (black_pix / (h_size[0] * h_size[1])) < 0.65 ):
                            shutil.copy(file_path, healthy_SAGIR)
                            healthy_counter +=1
    
    print("Total Unhealthy Images copied edema/fracture::", unhealthy_counter)
    print("Total Unhealthy Images having fracture::", unhealthy_frac_count)
    print("Total Unhealthy Images having edema::", unhealthy_edema_count)
    print("Total healthy Images copied::", healthy_counter)

In [87]:
# prepare the data
prepareData(imgName, img)

Total unique unhealty images are :::: 1016
Total  unique Images that has edema are ::: 1015
Total  unique Images that has fracture are ::: 325
Start of image seggregation to respective folders......
Total Unhealthy Images copied edema/fracture:: 861
Total Unhealthy Images having fracture:: 283
Total Unhealthy Images having edema:: 860
Total healthy Images copied:: 411


In [64]:
# import os

# def delete_ds_store_files(folder_path):
#     # Walk through all directories and files in the given folder path
#     for root, dirs, files in os.walk(folder_path):
#         for file in files:
#             # Check if the file is .DS_Store
#             if file == ".DS_Store":
#                 file_path = os.path.join(root, file)
#                 try:
#                     os.remove(file_path)
#                     print(f"Deleted: {file_path}")
#                 except Exception as e:
#                     print(f"Failed to delete {file_path}: {e}")

# # Specify the folder from which you want to delete .DS_Store files
# fp = "/Users/jiten/Masters/WorkPlace/MRI_Image/"

# # Call the function
# delete_ds_store_files(fp)